### PR TITLE
feat: create @waku/rln

### PR DESCRIPTION
**Pull request by weboko**
## Problem

`@waku/rln` is placed in a different repo with it's release pipeline

## Solution

Make it part of `js-waku`

## Notes

- Resolves https://github.com/waku-org/js-rln/issues/76

TODO:
- fixing some checks;
- double check pipelines

Overall review can be started.

### CODE REVIEW

**Review by adklempner — APPROVED** [review]

File: packages/rln/scripts/schema_validator_codegen.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import standaloneCode from "ajv/dist/standalone/index.js";
+
+import keystoreSchema from "./nwaku_keystore.json" assert { type: "json" };
+import credentialSchema from "./nwaku_keystore_credential.json" assert { type: "json" };
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = dirname(__filename);
+
+// Generate a function that validates the Keystore schema.
+// Write the function to a file in the source.
```
Comment (danisharora099) [inline packages/rln/scripts/schema_validator_codegen.js:17]: nit: what is the relevance of these comments?

File: packages/rln/src/codec.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
+import type {
```
Comment (danisharora099) [inline packages/rln/src/codec.ts:1]: nit: let's split this file into two: `Decoder` and `Encoder` -- it's quite long as is

File: packages/rln/src/contract/rln_contract.ts [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+    skip += size;
+  }
+}
+
+function ignoreErrors<T>(promise: Promise<T>, defaultValue: T): Promise<T> {
+  return promise.catch((err) => {
+    log(`Ignoring an error during query: ${err?.message}`);
+    return defaultValue;
+  });
+}
```
Comment (danisharora099) [inline packages/rln/src/contract/rln_contract.ts:355]: nit: let's move utils into a separate file separate from the main `RLNContract`

File: packages/rln/tsconfig.karma.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+  "extends": "./tsconfig.dev",
+  "compilerOptions": {
+    "noEmit": false
```
Comment (danisharora099) [inline packages/rln/tsconfig.karma.json:4]: why?

File: packages/rln/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  "compilerOptions": {
+    "incremental": true,
+    "target": "ES2022",
+    "moduleResolution": "Bundler",
+    "module": "ES2022",
+    "declaration": true,
+    "sourceMap": true,
+    "esModuleInterop": true /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */,
+    "resolveJsonModule": true /* Include modules imported with .json extension. */,
+    "strict": true /* Enable all strict type-checking options. */,
```
Comment (danisharora099) [inline packages/rln/tsconfig.json:11]: why are we rewriting this? we should extend from root

File: packages/rln/tsconfig.dev.json [hunk 6]
```diff
@@ -0,0 +1,8 @@
+{
+  "extends": "./tsconfig",
+  "compilerOptions": {
+    "module": "ESNext",
+    "moduleResolution": "Bundler",
+    "noEmit": true
+  }
```
Comment (danisharora099) [inline packages/rln/tsconfig.dev.json:7]: the original tsconfig already has the same compiler options, we don't need to rewrite

File: packages/rln/src/utils/metamask.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { ethers } from "ethers";
+
+export const extractMetaMaskSigner = async (): Promise<ethers.Signer> => {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const ethereum = (window as any).ethereum;
```
Comment (danisharora099) [inline packages/rln/src/utils/metamask.ts:5]: nit: typecasting as `unknown` could be better

File: packages/rln/src/utils/bytes.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    totalLength += arr.length;
+  }
+  const result = new Uint8Array(totalLength);
+  let offset = 0;
+  for (const arr of input) {
+    result.set(arr, offset);
+    offset += arr.length;
+  }
+  return result;
+}
```
Comment (danisharora099) [inline packages/rln/src/utils/bytes.ts:18]: we already have a `concat` utility in `@waku/utils`

other utils in this file could live there, if not already present

File: packages/rln/src/message.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    public rateLimitProof: IRateLimitProof | undefined
+  ) {}
+
+  public verify(roots: Uint8Array[]): boolean | undefined {
+    return this.rateLimitProof
+      ? this.rlnInstance.zerokit.verifyWithRoots(
+          this.rateLimitProof,
+          toRLNSignal(this.msg.contentTopic, this.msg),
+          ...roots
+        ) // this.rlnInstance.verifyRLNProof once issue status-im/nwaku#1248 is fixed
```
Comment (danisharora099) [inline packages/rln/src/message.ts:31]: https://github.com/waku-org/nwaku/issues/1248 is closed -- we should update this

File: packages/rln/src/message.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+      : undefined;
+  }
+
+  public verifyNoRoot(): boolean | undefined {
+    return this.rateLimitProof
+      ? this.rlnInstance.zerokit.verifyWithNoRoot(
+          this.rateLimitProof,
+          toRLNSignal(this.msg.contentTopic, this.msg)
+        ) // this.rlnInstance.verifyRLNProof once issue status-im/nwaku#1248 is fixed
```
Comment (danisharora099) [inline packages/rln/src/message.ts:40]: https://github.com/waku-org/nwaku/issues/1248 is closed -- we should update this

File: packages/rln/src/keystore/keystore.spec.ts [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+
+    expect(store.toObject()).to.deep.eq({
+      application: "waku-rln-relay",
+      appIdentifier: "01234567890abcdef",
+      version: "0.2",
+      credentials: {}
+    });
+  });
+
+  // TODO: add more thorow credentials testing
```
Comment (danisharora099) [inline packages/rln/src/keystore/keystore.spec.ts:109]: track in an issue?

**Review by danisharora099 — CHANGES_REQUESTED** [review]

File: packages/rln/src/keystore/keystore.spec.ts (hunk 11 above)
Comment (weboko) [reply packages/rln/src/keystore/keystore.spec.ts:109]: https://github.com/waku-org/js-waku/issues/1850

File: packages/rln/src/message.ts (hunk 9 above)
Comment (weboko) [reply packages/rln/src/message.ts:31]: good catch and I propose to have it as follow up as it can cause bags so needs additional attention

**Comment by weboko** [comment]
@danisharora099 thanks for the comments, I propose to have those that are actionable as follow ups

**Comment by weboko** [comment]
@danisharora099 thanks for the comments, I propose to have those that are actionable as follow ups

**Comment by danisharora099** [comment]
> @danisharora099 thanks for the comments, I propose to have those that are actionable as follow ups

Let's address the non-nit ones that are more important such as: not redefining `tsconfig` again, reusing the function from `@waku/utils` instead of redefining, etc.
If you prefer, I'm happy to make commits to address these.

Nits can be addressed in a followup PR!

**Comment by weboko** [comment]
Closing for now

**Pull request closed without merging**